projects/views.py

```diff
@@ -546,16 +590,60 @@ def program_edit(request, program_id=0):
         'status': 'success'
     })
 
+
 @login_required
 def program_share(request, program_id=0):
-    program = request.user.program_set.get(id=program_id)
-    emails = re.split('\s*,\s*', request.POST['share_users'].strip())
+    message = None
+    status = None
+    status_text = None
 
-    create_share(request, program, emails, 'program')
+    try:
+        # Verify all emails are in our user database
+        emails = re.split('\s*,\s*', request.POST['share_users'].strip())
+        users_not_found = []
+        users = []
+        req_user = None
+
+        try:
+            req_user = User.objects.get(id=request.user.id)
+        except ObjectDoesNotExist as e:
+            raise Exception("{} is not a user ID in this database!".format(str(request.user.id)))
+
+        for email in emails:
+            try:
+                user = User.objects.get(email=email)
+                users.append(user)
+            except ObjectDoesNotExist as e:
+                users_not_found.append(email)
+
+        # If any aren't found, warn the user and don't share
+        if len(users_not_found) > 0:
+            status_text = 'error'
+            # An actual error will close the modal, so this is a '200 error' i.e. 'ok request, but can't carry it out'
+            status = 200
+            message = 'The following user emails could not be found; please ask them to log into the site first: ' + ", ".join(users_not_found)
+
+        # Otherwise, share the program
+        else:
+            program = request.user.program_set.get(id=program_id)
+
+            create_share(request, program, emails, 'program')
+            status = 200
+            status_text = 'success'
+            message = 'Program ID {} has been successfully shared with the following user(s): {}'.format(str(program_id),", ".join(emails))
+
+    except Exception as e:
+        logger.error("[ERROR] While attempting to share program ID {}: ".format(str(program_id)))
+        logger.exception(e)
+        status_text = 'error'
+        status = 500
+        message = 'There was an error while attempting to share program ID {}.'.format(str(program_id))
```

Comment:
here's another spot where it looks like specific error messages raised earlier do not get to the user